website/blog/2023-4-10-incrementals-pt2.md

```diff
@@ -0,0 +1,538 @@
+indexing information in the next part.
+
+Data for items is pulled from M365 using the Graph API when kopia actually needs
+to upload the data to S3. By lazily fetching data, Corso avoids making Graph API
+requests if kopia already has a copy of that item’s data. It also reduces the
+memory footprint of Corso because only a few items’ data needs to be in memory
+at any given point in time.
+
+Going back to our running example, the folder merge operation for `folder1` would proceed as follows:
+
+```JSON
+// starting state
+global exclude items: [
+  super secret file1.txt (ID: 2),
+  tmp.log (ID: 3),
+  secret file.docx (ID: 5),
+  new secret file.docx (ID 13),
+]
+local exclude items: []
+
+// items from Collection
+super secret file1.txt (ID: 2) // uploaded, add to local exclude items
+
+// new state
+global exclude list: [
+  super secret file1.txt (ID: 2),
+  tmp.log (ID: 3),
+  secret file.docx (ID: 5),
+  new secret file.docx (ID 13),
+]
+local exclude items: [super secret file1.txt (ID: 2)]
+
+// items from base backup
+super secret file1.txt (ID: 2) // skipped because it's in local exclude items
+temp.log (ID: 3) // skipped because it's in the global exclude items
+```
+
+By the end of this part, Corso has also collected references for all items
+sourced from the base backup. The references that Corso tracks helps it map from
+the old location of the item to the new one as shown below.
+
+```JSON
+items from base backup: [
+  /root/folder3/static file.docx => /root/folder3/static file.docx
+  /root/folder4/plain file.txt => /root/folder1/folder4/plain file.txt
+  /root/folder4/folder5/update.log => /root/folder1/folder4/folder5/update.log
+]
+```
+
+## Part 6: Merging indexing information and persisting the Backup model
+
+The final thing Corso needs to do is merge the set of updated indexing
+information with indexing information from the base backup. Merging the two
+allows Corso to filter backup details and restore inputs down to just the
+selected items without having to traverse multiple backups to find all the
+information.
+
+To merge indexing information, Corso first loads the indexing information from
+the base backup. Then, it compares the entries in the base backup’s index with
+the set of references collected in the previous part. Entries that match are
+updated if needed (the item’s location in OneDrive may have changed) and
+added to the new set of indexing information for the current backup.
+
+Once the indexing information has been merged all data for the backup is
+complete except the BackupModel that acts as the root reference for the backup.
+The backup model summarizes what happened in the backup and contains references
+to the index information and the item data snapshot so they can be found later.
+The model itself is also stored in kopia so it’s data is encrypted as well. Once
+the BackupModel is added to kopia the Corso backup is considered complete and
+can be used for future restores.
+
```

Comment:
Needs a conclusion that covers a summary of what was covered in this blog post, pointers back to part 1, and why what we discussed is important.

Reply:
added a brief conclusion and link back to pt 1